File: examples/lib/array.cpp
#include "array.hpp"

namespace __array__ {

str *const_0;

str *__name__;

void * default_0;

/**
class array
*/

class_ *cl_array;

void *array::fromlist(list<__ss_int> *l) {
    for(int i=0; i<len(l); i++)
        unit->append(l->__getitem__(i));
    return NULL;
}

list<__ss_int> *array::tolist() {
    return unit->__slice__(0,0,0,0);
}

void *array::fromstring(str *s) {
    for(int i=0; i<len(s); i++)
        unit->append((unsigned char)(s->unit[i]));
    return NULL;
}

void *array::tofile(file *f) {
    f->write(tostring());
    return NULL;
}

str *array::tostring() {
    str *r = new str();
    for(int i=0; i<len(unit); i++)
        r->unit += ((unsigned char)(unit->__getitem__(i)));
    return r;
}

__ss_int array::__len__() {
    return len(unit);
}

array *array::__slice__(__ss_int x, __ss_int l, __ss_int u, __ss_int s) {
    return new array(new str("B"), unit->__slice__(x, l, u, s));
}

void *array::fromfile(file *f, __ss_int n) {
    str *r = f->read(n);
    fromstring(r);
    if(len(r) != n)
        throw new EOFError();
    return NULL;
}

void *array::__init__(str *flags, list<__ss_int> *arg) {
    unit = new list<__ss_int>();
    if(arg != NULL)
        fromlist(arg);
    return NULL;
}

void *array::__delete__(__ss_int x, __ss_int a, __ss_int b, __ss_int s) {
    unit->__delete__(x, a, b, s);
}

void *array::__setitem__(__ss_int i, __ss_int e) {
    unit->__setitem__(i, e);
}

str *array::__repr__() {
    if(len(unit))
        return __add_strs(3, new str("array('B', "), repr(unit), new str(")"));
    else
        return new str("array('B')");
}

void __init() {
    const_0 = new str("");

    __name__ = new str("array");

    cl_array = new class_("array", 29, 29);

    default_0 = NULL;
}

} // module namespace

File: shedskin/lib/array.py
# Copyright 2005-2011 Mark Dufour and contributors; License Expat (See LICENSE)


class array:
    def __init__(self, flag, it=None):
        self.typecode = ''
        self.itemsize = 0

    def __init_int__(self, flag, it=None):
        self.unit = 1
        self.__init__(flag)
    def __init_float__(self, flag, it=None):
        self.unit = 1.0
        self.__init__(flag)
    def __init_str__(self, flag, it=None):
        self.unit = ''
        self.__init__(flag)

    def tolist(self):
        return [self.unit]
    def tostring(self):
        return b''
    def tobytes(self):
        return b''

    def fromlist(self, l):
        pass
    def fromstring(self, s):
        pass
    def frombytes(self, s):
        pass

    def extend(self, it):
        pass

    def pop(self, i=-1):
        return self.unit
    def remove(self, e):
        pass

    def __getitem__(self, i):
        return self.unit
    def __setitem__(self, i, e):
        pass
    def __delitem__(self, i):
        pass

    def insert(self, i, e):
        pass
    def append(self, e):
        pass

    def tofile(self, f):
        pass
    def fromfile(self, f, n):
        pass

    def count(self, e):
        return 1
    def index(self, e):
        return 1
    def __contains__(self, e):
        return True

    def byteswap(self):
        pass
    def reverse(self):
        pass

    def __add__(self, b):
        return self
    def __mul__(self, b):
        return self
    def __with_int__(self): # int.__mul__ etc.
        return self

    def __len__(self):
        return 1

    def __iter__(self):
        return __iter(self.unit)

    def __slice__(self, x, lower, upper, step):
        return self
    def __delslice__(self, a, b):
        pass
    def __setslice__(self, x, lower, upper, step, r):
        pass
    def __delete__(self, x, a=1, b=1, s=1):
        pass

    def __copy__(self):
        return self
